xbmc/music/windows/GUIWindowVisualisation.cpp

```diff
@@ -240,3 +241,29 @@ void CGUIWindowVisualisation::FrameMove()
   }
   CGUIWindow::FrameMove();
 }
+
+void CGUIWindowVisualisation::ToggleOSD()
```

Comment:
Well, it also took me a while to understand this code (I hope I actually did). I'm afraid it will take some more words to explain things. :-/ So please take yourself the time to read through the following wall of text. ;-)

GUIWindowVisualisation and GUIWindowFullscreen to some extend are doing similar things, working almost the same, the first for music, the latter for video. Let me explain things on the music side:

GUIWindowVisualisation (among other things) is responsible for music related OSD, like title, start and end time and such, but not for player controls (explained later). The visibility of the information is managed by flipping a bool (GUIInfoManager::SetShowInfo) whenever a component of Kodi wants to show/hide the OSD. Confluence checks this bool ("Player.ShowInfo") in MusicVisualisation.xml and shows/hides stuff accordingly. 

GUIDialogMusicOSD implements and manages the OSD player controls, like start, stop, etc.

Thus, if you see an OSD with both information for the music and player controls, technically there is GUIWindowVisualisation having set Player.ShowInfo to true and a running DialogMusicOSD.

Putting the peaces together: GUIWindowVisualisation::ToggleOSD gets called to toggle the visibility state of the player controls only, not the visibility of the OSD music information. For this, it first looks whether the player controls are visible (=> DialogMusicOSD is running). If so, it closes the dialog to hide the player controls. If not it calls DialogMusicOSD::Open to show the player controls. 

The name ToggleOSD is misleading, if you ask me, as it actually manages only the visibility of the player controls. But naming was there for the video part and I reused/copied it for music. Maybe we should change the naming later in another PR.

The logic just described is not new with this PR. In the first place I just extended the music side to implement the toggle correctly, like the video side does (introduction of GUIWindowVisualisation::ToggleOSD, as a copy of GUIWindowFullscreen::ToggleOSD).

Then I added the logic for the confirm channel switch feature to GUIWindowVisualisation::ToggleOSD. Will explain this in another comment.